### PR TITLE
[jerrelllee] iP

### PR DESCRIPTION
**Pull request by jerrelllee**
# Fence
It's: 
- text-based
-  easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need is:
1. download it from [here](https://github.com/jerrelllee/ip/releases)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :wink:

> [!IMPORTANT]
> Some of the features are still being developed.

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

### CODE REVIEW

File: src/main/java/fence/Fence.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
+package fence;
```
Comment (dominic2412) [inline src/main/java/fence/Fence.java:1]: Your error handling with custom exceptions (e.g., NoSuchElementException, DateTimeParseException) is effective.
Consider adding more specific error messages to guide the user on what they did wrong. This can enhance the user experience.

File: src/main/java/fence/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
+package fence.parser;
+
+import fence.task.Deadline;
+import fence.task.Event;
+import fence.task.Task;
+import fence.task.Todo;
+import java.time.LocalDate;
+import java.util.NoSuchElementException;
+import java.util.StringTokenizer;
```
Comment (dominic2412) [inline src/main/java/fence/parser/Parser.java]: The use of StringTokenizer to parse commands is a clever choice. However, it might be a bit hard to follow for those who are not familiar with StringTokenizer.

File: src/main/java/fence/task/Task.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package fence.task;
```
Comment (dominic2412) [inline src/main/java/fence/task/Task.java:1]: The task management logic is clean and straightforward.

File: src/main/java/fence/ui/Ui.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
+package fence.ui;
```
Comment (dominic2412) [inline src/main/java/fence/ui/Ui.java:1]: The UI responses in the Ui class are user-friendly and straightforward, goodjob!

**Review by dominic2412 — COMMENTED** [review]
I really like how you've approached this task, especially the clear and logical flow of the Fence class. With a few tweaks and improvements, this could be even better! Keep up the good work, and I'm looking forward to seeing how you continue to refine this project.

File: src/main/java/fence/storage/Storage.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            for (int i = 0; i < tasks.getSize(); i++) {
+                fw.write(tasks.getTask(i).toTxt() + System.lineSeparator());
+            }
+            fw.close();
+        } catch (IOException e) {
+            System.out.println("Something went wrong: " + e.getMessage());
+        }
+    }
+
+    public ArrayList<Task> read(Parser parser) {
```
Comment (sayglenn) [inline src/main/java/fence/storage/Storage.java]: Perhaps this could be named better? e.g. readTasks

File: src/main/java/fence/parser/Parser.java
Comment (sayglenn) [inline src/main/java/fence/parser/Parser.java:1]: There's no javadocs in this file, which could slightly be confusing since it may not be easy discerning between two similarly named functions `ParseTxt` and `ParseInput`. Perhaps javadocs to discern the two would be suitable?

File: src/main/java/fence/task/Task.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
+    }
+
+    /**
+     * Sets the isDone field as true.
+     */
+    public void complete() {
```
Comment (sayglenn) [inline src/main/java/fence/task/Task.java:27]: Perhaps the method name could be more specific on what is being completed?

File: src/main/java/fence/task/Task.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        this.isDone = true;
+    }
+
+    /**
+     * Sets the isDone field as false.
+     */
+    public void undo() {
```
Comment (sayglenn) [inline src/main/java/fence/task/Task.java:34]: Similarly as for the complete method above, could the name be more specific?

File: src/test/java/fence/parser/ParserTest.java
Comment (sayglenn) [inline src/test/java/fence/parser/ParserTest.java:1]: The naming conventions for your tests are well done! I could understand the purposes of these tests clearly.

File: src/main/java/fence/ui/Ui.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        for (int i = 0; i < tasks.getSize(); i++) {
+            System.out.println(i + 1 + ". " + tasks.getTask(i));
+        }
+    }
+
+    /**
+     * Prints out the size of the task list.
+     * @param size Size of the task list.
+     */
+    public void count(int size) {
```
Comment (sayglenn) [inline src/main/java/fence/ui/Ui.java]: This could be more specific, perhaps countTasks will be more suitable?

**Review by sayglenn — COMMENTED** [review]
Some javadocs are missing for some classes which I assume would be added in due time. However, the general flow of the code looks good! I especially liked the descriptive titles of your test cases. Good job!

File: src/main/java/fence/tasklist/TaskList.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    public int getSize() {
+        return items.size();
+    }
+
+    /**
+     * Appends the given task to the end of the list.
+     * @param task Task to be added.
+     */
+    public void add(Task task) {
```
Comment (muller317) [inline src/main/java/fence/tasklist/TaskList.java:43]: Would it be better to name it as addTask instead?

**Review by muller317 — COMMENTED** [review]
Overall the code adhere to the coding standard well, the OOP structure is clear and well-organized.
